### PR TITLE
Add general purpose HeaderFilter

### PR DESCRIPTION
**Pull request by xz64**
Signed-off-by: S K (xz64) <xz64.xyz@gmail.com>

### CODE REVIEW

**Comment by joakime** [comment]
Is this related to #1607 ?

**Comment by joakime** [comment]
These filters would need configuration for include/exclude behavior.
Include/Exclude based on paths (at a minimum), perhaps MimeTypes too.
These filter should probably only act on GET requests too.

**Comment by xz64** [comment]
This is related to #1607 . Since the default servlet is better optimized for static content, I am implementing the header functionality with a filter. Based on your feedback, I will change it up a bit and add these parameters:

* includePattern - java regex for what urls are included
* excludePattern - java regex for what urls are excluded (if a url is both included and excluded, then it will be excluded)
* mimeTypes - csv string of mime types to which the header should be applied
* methods - csv string of http methods to apply the filter on
* headers - csv of what headers to set. The syntax would be like this (in XML):

```
<init-param>
    <param-name>headers</param-name>
    <param-value>
       set X-Frame-Options: DENY,
       setDate Expires: 31540000000,
       addDate Date: 0,
       "add Cache-Control: no-cache, no-store, must-revalidate"
    </param-value>
</init-param>
```
set will call response.setHeader, addDate will call response.addDateHeader etc
setDate and addDate will add the value to the current system time before setting the header.

**Comment by joakime** [comment]
Please see GzipHandler for some example behaviors for Include/Exclude logic.

https://github.com/eclipse/jetty.project/blob/jetty-9.4.x/jetty-server/src/main/java/org/eclipse/jetty/server/handler/gzip/GzipHandler.java

**Comment by xz64** [comment]
I have updated the commit with the changes suggested. Let me know if you have any additional feedback.

**Comment by joakime** [comment]
I have to say, this is the most complete PR I've seen to date.

- [x] New functionality.
- [x] Test cases.
- [x] Documentation.

File: jetty-servlets/src/main/java/org/eclipse/jetty/servlets/IncludeExcludeBasedFilter.java [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+ * <li>If the spec starts with <code>'*.'</code> the spec is assumed to be a Servlet url-pattern rules path spec for a suffix based match.</li>
+ * <li>All other syntaxes are unsupported.</li>
+ * </ul>
+ * <p>
+ * CSVs are parsed with {@link StringUtil#csvSplit(String)}
+ *
+ * @see PathSpecSet
+ * @see IncludeExcludeSet
+ */
+public abstract class IncludeExcludeBasedFilter implements Filter
```
Comment (joakime) [inline jetty-servlets/src/main/java/org/eclipse/jetty/servlets/IncludeExcludeBasedFilter.java]: Great idea! 👍

**Pull request merged**